### ISSUE TITLE
大佬你好，我想问一下你这个用mobilenet训练出来的模型有比cnn的小吗？小了多少呢？是不是和理论上的差不多？（8到9倍）

### ISSUE THREAD

**Issue by llllearner** [GuodongQi/yolo3_tensorflow#6]

**Comment by GuodongQi** [comment]
是的，大约是原来的1/8

**Comment by EEeEhh** [comment]
我想问一个问题，就是我把原来yolov3的实现中的普通卷积层改成mobilenet中的深度可分离卷积层，为啥跑出来的模型一点变化都没有啊！真的是好气！大佬可以帮忙初步分析一下嘛

**Comment by EEeEhh** [comment]
`def convolutional(input_data, filters_shape, trainable, name, downsample=False, activate=True, bn=True):
    with tf.variable_scope(name):
        if downsample:
            pad_h, pad_w = (filters_shape[0] - 2) // 2 + 1, (filters_shape[1] - 2) // 2 + 1
            paddings = tf.constant([[0, 0], [pad_h, pad_h], [pad_w, pad_w], [0, 0]])
            input_data = tf.pad(input_data, paddings, 'CONSTANT')
            strides = (1, 2, 2, 1)
            padding = 'VALID'
        else:
            strides = (1, 1, 1, 1)
            padding = "SAME"

        weight = tf.get_variable(name='weight', dtype=tf.float32, trainable=True,
                                 shape=filters_shape, initializer=tf.random_normal_initializer(stddev=0.01))
        conv = tf.nn.conv2d(input=input_data, filter=weight, strides=strides, padding=padding)

        if bn:
            conv = tf.layers.batch_normalization(conv, beta_initializer=tf.zeros_initializer(),
                                                 gamma_initializer=tf.ones_initializer(),
                                                 moving_mean_initializer=tf.zeros_initializer(),
                                                 moving_variance_initializer=tf.ones_initializer(), training=trainable)
        else:
            bias = tf.get_variable(name='bias', shape=filters_shape[-1], trainable=True,
                                   dtype=tf.float32, initializer=tf.constant_initializer(0.0))
            conv = tf.nn.bias_add(conv, bias)

        if activate:
            conv = tf.nn.leaky_relu(conv, alpha=0.1)

    return conv


def conv_block(input_data, filters_shape, trainable, name, net_type, downsample=False, activate=True, bn=True):
    with tf.variable_scope('' + name):
        if net_type == 'cnn':
            with tf.variable_scope('cnn'):
                if downsample:  # refer to "https://github.com/qqwweee/keras-yolo3/issues/8"
                    pad_h, pad_w = (filters_shape[0] - 2) // 2 + 1, (filters_shape[1] - 2) // 2 + 1
                    paddings = tf.constant([[0, 0], [pad_h, pad_h], [pad_w, pad_w], [0, 0]])
                    input_data = tf.pad(input_data, paddings=paddings)
                    stride = (1, 2, 2, 1)
                    padding = 'VALID'
                else:
                    stride = (1, 1, 1, 1)
                    padding = 'SAME'
                weight = tf.get_variable(name='weights', shape=filters_shape, initializer=rand_initializer)
                x = tf.nn.conv2d(input_data, weight, strides=stride, padding=padding)

                if bn:
                    x = tf.layers.batch_normalization(x, training=trainable)
                else:
                    bias = tf.get_variable(name='bias', shape=filters_shape[-1], initializer=tf.constant_initializer(0))
                    x = tf.nn.bias_add(x, bias)
                if activate:
                    x = tf.nn.leaky_relu(x, 0.1)

        elif net_type == 'mobilenetv1':
            with tf.variable_scope('depthwise'):
                if downsample:
                    stride = (1, 2, 2, 1)
                else:
                    stride = (1, 1, 1, 1)
                depthwise_weight = tf.get_variable(name='weights', shape=([filters_shape[0], filters_shape[1], filters_shape[2], 1]),
                                                   initializer=rand_initializer)  #
                x = tf.nn.depthwise_conv2d(input_data, depthwise_weight, strides=stride, padding='SAME')
                x = tf.layers.batch_normalization(x, training=trainable)
                x = tf.nn.relu6(x)

            with tf.variable_scope('pointwise'):
                pointwise_weight = tf.get_variable(name='weights', shape=filters_shape, initializer=rand_initializer)
                x = tf.nn.conv2d(x, pointwise_weight, [1, 1, 1, 1], 'SAME')

                if bn:
                    x = tf.layers.batch_normalization(x, training=trainable)
                else:
                    bias = tf.get_variable(name='bias',shape=filters_shape[-1], initializer=tf.constant_initializer(0))

                    x = tf.nn.bias_add(x, bias)
                if activate:
                    x = tf.nn.relu6(x)
        else:
            raise Exception('net type is error, please check')
    return x`

**Comment by EEeEhh** [comment]
```
def convolutional(input_data, filters_shape, trainable, name, downsample=False, activate=True, bn=True):
    with tf.variable_scope(name):
        if downsample:
            pad_h, pad_w = (filters_shape[0] - 2) // 2 + 1, (filters_shape[1] - 2) // 2 + 1
            paddings = tf.constant([[0, 0], [pad_h, pad_h], [pad_w, pad_w], [0, 0]])
            input_data = tf.pad(input_data, paddings, 'CONSTANT')
            strides = (1, 2, 2, 1)
            padding = 'VALID'
        else:
            strides = (1, 1, 1, 1)
            padding = "SAME"

        weight = tf.get_variable(name='weight', dtype=tf.float32, trainable=True,
                                 shape=filters_shape, initializer=tf.random_normal_initializer(stddev=0.01))
        conv = tf.nn.conv2d(input=input_data, filter=weight, strides=strides, padding=padding)

        if bn:
            conv = tf.layers.batch_normalization(conv, beta_initializer=tf.zeros_initializer(),
                                                 gamma_initializer=tf.ones_initializer(),
                                                 moving_mean_initializer=tf.zeros_initializer(),
                                                 moving_variance_initializer=tf.ones_initializer(), training=trainable)
        else:
            bias = tf.get_variable(name='bias', shape=filters_shape[-1], trainable=True,
                                   dtype=tf.float32, initializer=tf.constant_initializer(0.0))
            conv = tf.nn.bias_add(conv, bias)

        if activate:
            conv = tf.nn.leaky_relu(conv, alpha=0.1)

    return conv


def conv_block(input_data, filters_shape, trainable, name, net_type, downsample=False, activate=True, bn=True):
    with tf.variable_scope('' + name):
        if net_type == 'cnn':
            with tf.variable_scope('cnn'):
                if downsample:  # refer to "https://github.com/qqwweee/keras-yolo3/issues/8"
                    pad_h, pad_w = (filters_shape[0] - 2) // 2 + 1, (filters_shape[1] - 2) // 2 + 1
                    paddings = tf.constant([[0, 0], [pad_h, pad_h], [pad_w, pad_w], [0, 0]])
                    input_data = tf.pad(input_data, paddings=paddings)
                    stride = (1, 2, 2, 1)
                    padding = 'VALID'
                else:
                    stride = (1, 1, 1, 1)
                    padding = 'SAME'
                weight = tf.get_variable(name='weights', shape=filters_shape, initializer=rand_initializer)
                x = tf.nn.conv2d(input_data, weight, strides=stride, padding=padding)

                if bn:
                    x = tf.layers.batch_normalization(x, training=trainable)
                else:
                    bias = tf.get_variable(name='bias', shape=filters_shape[-1], initializer=tf.constant_initializer(0))
                    x = tf.nn.bias_add(x, bias)
                if activate:
                    x = tf.nn.leaky_relu(x, 0.1)

        elif net_type == 'mobilenetv1':
            with tf.variable_scope('depthwise'):
                if downsample:
                    stride = (1, 2, 2, 1)
                else:
                    stride = (1, 1, 1, 1)
                depthwise_weight = tf.get_variable(name='weights', shape=([filters_shape[0], filters_shape[1], filters_shape[2], 1]),
                                                   initializer=rand_initializer)  #
                x = tf.nn.depthwise_conv2d(input_data, depthwise_weight, strides=stride, padding='SAME')
                x = tf.layers.batch_normalization(x, training=trainable)
                x = tf.nn.relu6(x)

            with tf.variable_scope('pointwise'):
                pointwise_weight = tf.get_variable(name='weights', shape=filters_shape, initializer=rand_initializer)
                x = tf.nn.conv2d(x, pointwise_weight, [1, 1, 1, 1], 'SAME')

                if bn:
                    x = tf.layers.batch_normalization(x, training=trainable)
                else:
                    bias = tf.get_variable(name='bias',shape=filters_shape[-1], initializer=tf.constant_initializer(0))

                    x = tf.nn.bias_add(x, bias)
                if activate:
                    x = tf.nn.relu6(x)
        else:
            raise Exception('net type is error, please check')
    return x
```

**Comment by llllearner** [comment]
> 是的，大约是原来的1/8

为啥我跑出来的只有1/4左右啊，好气！

**Comment by GuodongQi** [comment]
这个要看你自己是如何搭建模型的了。比如前面特征提取层（darknet）采取mobilenet结构，而最后几层采取cnn结构时，也不严格是原来的1/8

**Comment by llllearner** [comment]
> 这个要看你自己是如何搭建模型的了。比如前面特征提取层（darknet）采取mobilenet结构，而最后几层采取cnn结构时，也不严格是原来的1/8

是的，谢谢。

**Comment by Wangfeng2394** [comment]
我训练的 cnn是250m左右           mobilenet是30m 差不多是8到9倍
